Allow writing custom test runners

algernon opened the pull request:
This introduces a new `.get_tests()` method for `Module` (and a wrapper around it for `Compiled`), which returns an iterator over `codegen::TestCase`, a helper struct that encapsulates the test's name, and the compiled function.

This iterator can be used to implement custom runners that filter tests, run them with independent contexts, or provide a different kind of output than the built in `run_tests()` method does, and so on.

The existing `run_tests()` method was also changed to use `.get_tests()`.

Test cases to exercise the new method are included.

tertsdiepraam commented:
Hmmm, when you told me about this issue, I was imagining a slightly different solution where I don't write anything but just give an iterator of test results that you can then write out (as text or JSON or whatever you want).

I guess my question for you is whether you think that would be worth the effort? Would you like structured output?

algernon commented:
I'm halfway through an implementation, and... it feels nice to use so far. Will push what I have in a bit.

algernon commented:
Sneak peek:

```rust
let mut all_passed = true;
let mut n = 0;
for test in compiled.get_tests::<IocaineContext>() {
    n += 1;
    let message;

    if test.run(&mut context) == Ok(()) {
        message = "ok";
    } else {
        message = "not ok";
        all_passed = false;
    }
    println!("{message} {n} - {}", test.name());
}
println!("1..{n}");
```

This produces a simple TAP output:

```
ok 1 - pkg.nam_shub_of_enki.detect_faked_agents_regexp_mozilla6
ok 2 - pkg.nam_shub_of_enki.detect_archivers
ok 3 - pkg.nam_shub_of_enki.detect_commercial_scrapers
not ok 4 - pkg.nam_shub_of_enki.classification_good_bots_feed_readers
ok 5 - pkg.nam_shub_of_enki.classification_good_bots_fedi_software_by_referrer
ok 6 - pkg.nam_shub_of_enki.classification_good_bots_fedi_software_by_path
ok 7 - pkg.nam_shub_of_enki.classification_good_bots_fedi_software_by_agent
ok 8 - pkg.nam_shub_of_enki.detect_demo
ok 9 - pkg.nam_shub_of_enki.detect_ancient_agents_regexp
ok 10 - pkg.nam_shub_of_enki.detect_cgi_bin_trap
ok 11 - pkg.nam_shub_of_enki.detect_x_firefox_ai
ok 12 - pkg.nam_shub_of_enki.detect_ai_robots_txt
ok 13 - pkg.nam_shub_of_enki.detect_headless_browsers
ok 14 - pkg.nam_shub_of_enki.detect_big_tech
ok 15 - pkg.nam_shub_of_enki.detect_ancient_agents_exception
ok 16 - pkg.nam_shub_of_enki.classification_good_bots_comm_software_by_agent
ok 17 - pkg.nam_shub_of_enki.classification_good_bots_tools_and_services
ok 18 - pkg.nam_shub_of_enki.detect_ancient_agents_pattern
ok 19 - pkg.nam_shub_of_enki.detect_faked_agents_regexp_mozilla45
ok 20 - pkg.nam_shub_of_enki.detect_anti_robots_txt
1..20
```

tertsdiepraam commented:
I was trying too, but you seem to be doing better. I was in lifetime hell 😄

algernon commented:
Pushed a WIP, the code looks okay-ish, but I want to write a test that exercises `.get_tests()` directly, not only via `.run_tests()`. And I'd love to get rid of the `.collect()` in `.get_tests()` too... although, it only collects names, so it's not that big of a deal. Still irks me a bit.

tertsdiepraam commented on src/codegen/mod.rs:
```diff
@@ -1160,35 +1177,57 @@ impl Module {
+                Ok(()) => {
                     successes += 1;
-                    println!("\x1B[92mok\x1B[m");
+                    writeln!(&mut w, "\x1B[92mok\x1B[m").map_err(|_| ())?;
```
Maybe make the return type `io::Result<Result<(), ()>>` and then just `?` these

algernon replied:
Another option would be to drop this part, keep using stdout, and if anyone wants to write to some other place: there's `.get_tests()`. The wrapping around it is - imo - small enough to be reimplemented.

I mean, there are ANSI colorcodes embedded here, which don't make much sense, unless the output is std{out,err}.

What do you think?

algernon replied:
If the need for a `run_tests_with_writer()` arises, it can always be added later, in a way that satisfies *those* needs better. I mean, it's not a particularly flexible function in the shape I implemented it... it was Just Enough™ to satisfy a need. As it happens, `.get_tests()` satisfies that need much better.

I'll remove the function then, and will change the title & description of the PR too, once I'm done with the rest of the cleanups.

tertsdiepraam commented on src/codegen/mod.rs:
```diff
@@ -1160,35 +1177,57 @@ impl Module {
+    pub fn run_tests_with_writer<Ctx: 'static>(
+        &mut self,
+        mut ctx: Ctx,
+        mut w: &mut impl std::io::Write,
```
I think taking `mut w: impl std::io::Write` should be enough. You're taking references elsewhere.

tertsdiepraam reviewed (commented):
Excellent! Some small comments, but this is a good way to do it

algernon commented on src/codegen/mod.rs:
```diff
@@ -1165,31 +1177,47 @@ impl Module {
             .map(Clone::clone)
             .collect();
 
+        tests.into_iter().map(|name| TestCase {
```
Unfortunately, the `collect()` is necessary. That's the thing that lets us avoid lifetime hell, because it unbinds the testcase name list from the lifetime of `self`.

algernon commented:
There we go! I removed the `.run_tests_with_writer()` method, leaving only `.get_tests()`, and the adjusted `.run_tests()`.

I moved the `TestCase` struct into its own module, because I wanted to make it mostly opaque, so that nothing relies on its current members, not even by accident, and the only way to work with it, is through its public functions.

The PR includes two simple tests, both exercising `.get_tests()`: one to collect & run tests without `.run_tests()`, and another to verify that nothing blows up if there are no tests present.

tertsdiepraam reviewed (approved):
Happy to accept this, but maybe the module should be `testing` and contain all the testing code?

algernon commented:
> Nice! Happy to accept this, but maybe the module should be `testing` and contain all the testing code?

I can do that. Will push an update soonish (an hour or so, am in the middle of something else at the moment).

algernon commented:
Turns out, this was easier than anticipated. I kept it as a separate commit for reviewing purposes, and because if something needs to be changed, or reverted, it's easier that way.
